File: Sources/HoverPocket/Services/GoogleOAuthService.swift
import AppKit
import CryptoKit
import Foundation
import Security

struct GoogleOAuthConfiguration: Equatable, Sendable {
    let googleSignInClientID: String?
    let legacyClientID: String?
    let legacyClientSecret: String?

    static var current: GoogleOAuthConfiguration? {
        let googleSignInClientID = Self.trimmedInfoValue("GIDClientID")
        let legacyClientID = Self.trimmedInfoValue("GoogleOAuthClientID")
        guard googleSignInClientID != nil || legacyClientID != nil else {
            return nil
        }

        return GoogleOAuthConfiguration(
            googleSignInClientID: googleSignInClientID,
            legacyClientID: legacyClientID,
            legacyClientSecret: Self.trimmedInfoValue("GoogleOAuthClientSecret")
        )
    }

    var oauthClientID: String? {
        googleSignInClientID ?? legacyClientID
    }

    var usesNativeAppOAuth: Bool {
        googleSignInClientID != nil
    }

    var nativeCallbackScheme: String? {
        guard let googleSignInClientID else { return nil }
        return Self.reversedClientID(googleSignInClientID)
    }

    var nativeRedirectURI: String? {
        guard let nativeCallbackScheme else { return nil }
        return "\(nativeCallbackScheme):/oauth2redirect/google"
    }

    private static func trimmedInfoValue(_ key: String) -> String? {
        let rawValue = Bundle.main.object(forInfoDictionaryKey: key) as? String
        let value = rawValue?.trimmingCharacters(in: .whitespacesAndNewlines)
        return value?.isEmpty == false ? value : nil
    }

    private static func reversedClientID(_ clientID: String) -> String {
        clientID
            .split(separator: ".")
            .reversed()
            .joined(separator: ".")
    }
}

struct GoogleOAuthToken: Codable, Equatable, Sendable {
    let accessToken: String
    let expiresAt: Date
    let grantedScopes: [String]

    var isFresh: Bool {
        expiresAt.timeIntervalSinceNow > 60
    }
}

enum GoogleOAuthStoredCredentialStatus: Equatable {
    case missing
    case needsReconnect
    case ready
}

enum GoogleOAuthError: LocalizedError {
    case missingConfiguration
    case browserOpenFailed
    case missingAuthorizationCode
    case stateMismatch
    case userDenied(String)
    case missingRefreshToken
    case insufficientScopes
    case storedCredentialRequiresReconnect
    case tokenEndpointFailed(String)
    case timedOut

    var errorDescription: String? {
        switch self {
        case .missingConfiguration:
            return "Google OAuth client ID is not configured."
        case .browserOpenFailed:
            return "Could not open the Google sign-in page."
        case .missingAuthorizationCode:
            return "Google did not return an authorization code."
        case .stateMismatch:
            return "Google sign-in state validation failed."
        case .userDenied(let message):
            return message
        case .missingRefreshToken:
            return "Google did not return a refresh token."
        case .insufficientScopes:
            return "Reconnect Google Calendar to allow event editing."
        case .storedCredentialRequiresReconnect:
            return "Reconnect Google Calendar to continue."
        case .tokenEndpointFailed(let message):
            return message
        case .timedOut:
            return "Google sign-in timed out."
        }
    }

    var requiresReconnect: Bool {
        switch self {
        case .missingRefreshToken, .insufficientScopes, .storedCredentialRequiresReconnect:
            return true
        case .missingConfiguration, .browserOpenFailed, .missingAuthorizationCode, .stateMismatch, .userDenied, .tokenEndpointFailed, .timedOut:
            return false
        }
    }
}

final class GoogleOAuthService: @unchecked Sendable {
    static let calendarScopes = [
        "https://www.googleapis.com/auth/calendar.calendarlist.readonly",
        "https://www.googleapis.com/auth/calendar.events"
    ]

    private let keychain = GoogleOAuthKeychainStore()
    private let tokenLock = NSLock()
    private var currentToken: GoogleOAuthToken?

    var isConfigured: Bool {
        GoogleOAuthConfiguration.current != nil
    }

    func hasStoredCredential() -> Bool {
        storedCredentialStatus() != .missing
    }

    func hasRequiredCalendarCredential() -> Bool {
        storedCredentialStatus() == .ready
    }

    func storedCredentialStatus() -> GoogleOAuthStoredCredentialStatus {
        let credential: GoogleOAuthStoredCredential
        do {
            guard let loadedCredential = try keychain.load() else {
                return .missing
            }
            credential = loadedCredential
        } catch {
            return .needsReconnect
        }
        return Self.hasRequiredCalendarScopes(credential.grantedScopes) ? .ready : .needsReconnect
    }

    func removeStoredCredential() {
        setCurrentToken(nil)
        keychain.delete()
    }

    func signIn() async throws {
        guard let configuration = GoogleOAuthConfiguration.current else {
            throw GoogleOAuthError.missingConfiguration
        }

        if configuration.usesNativeAppOAuth {
            try await signInWithNativeAppOAuth(configuration: configuration)
            return
        }

        try await signInWithLegacyOAuth(configuration: configuration)
    }

    func signOut() {
        let refreshToken = try? keychain.load()?.refreshToken
        setCurrentToken(nil)
        keychain.delete()
        if let refreshToken {
            Task.detached(priority: .utility) {
                try? await Self.revokeToken(refreshToken)
            }
        }
    }

    func accessToken() async throws -> String {
        try await accessToken(forceRefresh: false)
    }

    func accessToken(forceRefresh: Bool) async throws -> String {
        guard GoogleOAuthConfiguration.current != nil else {
            throw GoogleOAuthError.missingConfiguration
        }
        return try await legacyAccessToken(forceRefresh: forceRefresh)
    }

    private func signInWithLegacyOAuth(configuration: GoogleOAuthConfiguration) async throws {
        guard configuration.legacyClientID != nil else {
            throw GoogleOAuthError.missingConfiguration
        }

        let receiver = try LoopbackOAuthReceiver()
        let state = try randomBase64URL(byteCount: 32)
        let verifier = try randomBase64URL(byteCount: 64)
        let challenge = codeChallenge(for: verifier)
        let authURL = try authorizationURL(
            configuration: configuration,
            redirectURI: receiver.redirectURI,
            state: state,
            codeChallenge: challenge
        )

        let opened = await openAuthorizationURL(authURL)
        guard opened else {
            receiver.cancel()
            throw GoogleOAuthError.browserOpenFailed
        }

        let callback = try await waitForCallback(receiver)
        if let error = callback.error {
            throw GoogleOAuthError.userDenied(error)
        }
        guard callback.state == state else {
            throw GoogleOAuthError.stateMismatch
        }
        guard let code = callback.code, !code.isEmpty else {
            throw GoogleOAuthError.missingAuthorizationCode
        }

        let response = try await exchangeAuthorizationCode(
            code,
            verifier: verifier,
            redirectURI: receiver.redirectURI,
            configuration: configuration
        )
        guard let refreshToken = response.refreshToken, !refreshToken.isEmpty else {
            throw GoogleOAuthError.missingRefreshToken
        }

        let scopes = response.scope?.split(separator: " ").map(String.init) ?? Self.calendarScopes
        try keychain.save(
            GoogleOAuthStoredCredential(refreshToken: refreshToken, grantedScopes: scopes)
        )
        setCurrentToken(GoogleOAuthToken(
            accessToken: response.accessToken,
            expiresAt: Date().addingTimeInterval(TimeInterval(response.expiresIn)),
            grantedScopes: scopes
        ))
    }

    private func legacyAccessToken(forceRefresh: Bool) async throws -> String {
        if !forceRefresh, let currentToken = readCurrentToken(), currentToken.isFresh {
            guard Self.hasRequiredCalendarScopes(currentToken.grantedScopes) else {
                throw GoogleOAuthError.insufficientScopes
            }
            return currentToken.accessToken
        }

        let credential: GoogleOAuthStoredCredential
        do {
            guard let loadedCredential = try keychain.load() else {
                throw GoogleOAuthError.missingRefreshToken
            }
            credential = loadedCredential
        } catch let error as GoogleOAuthError {
            throw error
        } catch {
            removeStoredCredential()
            throw GoogleOAuthError.storedCredentialRequiresReconnect
        }
        guard Self.hasRequiredCalendarScopes(credential.grantedScopes) else {
            throw GoogleOAuthError.insufficientScopes
        }

        let refreshed: GoogleOAuthTokenResponse
        do {
            refreshed = try await refreshAccessToken(refreshToken: credential.refreshToken)
        } catch {
            if Self.shouldRemoveStoredCredential(after: error) {
                removeStoredCredential()
            }
            throw error
        }
        let scopes = refreshed.scope?.split(separator: " ").map(String.init) ?? credential.grantedScopes
        guard Self.hasRequiredCalendarScopes(scopes) else {
            throw GoogleOAuthError.insufficientScopes
        }
        let token = GoogleOAuthToken(
            accessToken: refreshed.accessToken,
            expiresAt: Date().addingTimeInterval(TimeInterval(refreshed.expiresIn)),
            grantedScopes: scopes
        )
        setCurrentToken(token)
        return token.accessToken
    }

    private static func shouldRemoveStoredCredential(after error: Error) -> Bool {
        (error as? GoogleOAuthError)?.requiresReconnect == true
    }

    private func readCurrentToken() -> GoogleOAuthToken? {
        tokenLock.lock()
        defer { tokenLock.unlock() }
        return currentToken
    }

    private func setCurrentToken(_ token: GoogleOAuthToken?) {
        tokenLock.lock()
        currentToken = token
        tokenLock.unlock()
    }

    private func waitForCallback(_ receiver: LoopbackOAuthReceiver) async throws -> OAuthCallback {
        try await withThrowingTaskGroup(of: OAuthCallback.self) { group in
            group.addTask {
                try await receiver.waitForCallback()
            }
            group.addTask {
                try await Task.sleep(for: .seconds(180))
                receiver.cancel()
                throw GoogleOAuthError.timedOut
            }

            guard let result = try await group.next() else {
                throw GoogleOAuthError.timedOut
            }
            group.cancelAll()
            return result
        }
    }

    private func authorizationURL(
        configuration: GoogleOAuthConfiguration,
        redirectURI: String,
        state: String,
        codeChallenge: String
    ) throws -> URL {
        guard let clientID = configuration.oauthClientID else {
            throw GoogleOAuthError.missingConfiguration
        }

        var components = URLComponents(string: "https://accounts.google.com/o/oauth2/v2/auth")!
        components.queryItems = [
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "redirect_uri", value: redirectURI),
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "scope", value: Self.calendarScopes.joined(separator: " ")),
            URLQueryItem(name: "access_type", value: "offline"),
            URLQueryItem(name: "include_granted_scopes", value: "true"),
            URLQueryItem(name: "prompt", value: "consent"),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "code_challenge", value: codeChallenge),
            URLQueryItem(name: "code_challenge_method", value: "S256")
        ]
        return components.url!
    }

    @MainActor
    private func openAuthorizationURL(_ url: URL) async -> Bool {
        NSWorkspace.shared.open(url)
    }

    private func signInWithNativeAppOAuth(configuration: GoogleOAuthConfiguration) async throws {
        guard
            let redirectURI = configuration.nativeRedirectURI,
            let callbackScheme = configuration.nativeCallbackScheme
        else {
            throw GoogleOAuthError.missingConfiguration
        }

        let state = try randomBase64URL(byteCount: 32)
        let verifier = try randomBase64URL(byteCount: 64)
        let challenge = codeChallenge(for: verifier)
        let authURL = try authorizationURL(
            configuration: configuration,
            redirectURI: redirectURI,
            state: state,
            codeChallenge: challenge
        )

        let receiver = OAuthURLCallbackReceiver(callbackScheme: callbackScheme)
        let opened = await openAuthorizationURL(authURL)
        guard opened else {
            receiver.cancel()
            throw GoogleOAuthError.browserOpenFailed
        }

        let callback = try await waitForCallback(receiver)
        if let error = callback.error {
            throw GoogleOAuthError.userDenied(error)
        }
        guard callback.state == state else {
            throw GoogleOAuthError.stateMismatch
        }
        guard let code = callback.code, !code.isEmpty else {
            throw GoogleOAuthError.missingAuthorizationCode
        }

        let response = try await exchangeAuthorizationCode(
            code,
            verifier: verifier,
            redirectURI: redirectURI,
            configuration: configuration
        )
        guard let refreshToken = response.refreshToken, !refreshToken.isEmpty else {
            throw GoogleOAuthError.missingRefreshToken
        }

        let scopes = response.scope?.split(separator: " ").map(String.init) ?? Self.calendarScopes
        try keychain.save(
            GoogleOAuthStoredCredential(refreshToken: refreshToken, grantedScopes: scopes)
        )
        setCurrentToken(GoogleOAuthToken(
            accessToken: response.accessToken,
            expiresAt: Date().addingTimeInterval(TimeInterval(response.expiresIn)),
            grantedScopes: scopes
        ))
    }

    private func waitForCallback(_ receiver: OAuthURLCallbackReceiver) async throws -> OAuthCallback {
        try await withThrowingTaskGroup(of: OAuthCallback.self) { group in
            group.addTask {
                try await receiver.waitForCallback()
            }
            group.addTask {
                try await Task.sleep(for: .seconds(180))
                throw GoogleOAuthError.timedOut
            }

            do {
                guard let result = try await group.next() else {
                    receiver.cancel()
                    throw GoogleOAuthError.timedOut
                }
                group.cancelAll()
                return result
            } catch GoogleOAuthError.timedOut {
                receiver.cancel()
                group.cancelAll()
                throw GoogleOAuthError.timedOut
            } catch {
                group.cancelAll()
                throw error
            }
        }
    }

    private func exchangeAuthorizationCode(
        _ code: String,
        verifier: String,
        redirectURI: String,
        configuration: GoogleOAuthConfiguration
    ) async throws -> GoogleOAuthTokenResponse {
        guard let clientID = configuration.oauthClientID else {
            throw GoogleOAuthError.missingConfiguration
        }

        var form = [
            "client_id": clientID,
            "code": code,
            "code_verifier": verifier,
            "grant_type": "authorization_code",
            "redirect_uri": redirectURI
        ]
        if !configuration.usesNativeAppOAuth, let clientSecret = configuration.legacyClientSecret {
            form["client_secret"] = clientSecret
        }
        return try await postTokenRequest(form: form)
    }

    private func refreshAccessToken(refreshToken: String) async throws -> GoogleOAuthTokenResponse {
        guard let configuration = GoogleOAuthConfiguration.current,
              let clientID = configuration.oauthClientID else {
            throw GoogleOAuthError.missingConfiguration
        }
        var form = [
            "client_id": clientID,
            "grant_type": "refresh_token",
            "refresh_token": refreshToken
        ]
        if !configuration.usesNativeAppOAuth, let clientSecret = configuration.legacyClientSecret {
            form["client_secret"] = clientSecret
        }
        return try await postTokenRequest(form: form, treatsStoredCredentialFailureAsReconnect: true)
    }

    private static func revokeToken(_ token: String) async throws {
        var request = URLRequest(url: URL(string: "https://oauth2.googleapis.com/revoke")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = "token=\(formEscape(token))".data(using: .utf8)

        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            return
        }
    }

    private func postTokenRequest(
        form: [String: String],
        treatsStoredCredentialFailureAsReconnect: Bool = false
    ) async throws -> GoogleOAuthTokenResponse {
        var request = URLRequest(url: URL(string: "https://oauth2.googleapis.com/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = percentEncodedForm(form).data(using: .utf8)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let responseError = try? JSONDecoder().decode(GoogleOAuthErrorResponse.self, from: data)
            if treatsStoredCredentialFailureAsReconnect,
               responseError?.requiresStoredCredentialReconnect == true {
                throw GoogleOAuthError.storedCredentialRequiresReconnect
            }
            throw GoogleOAuthError.tokenEndpointFailed(responseError?.safeDescription ?? "Google token request failed.")
        }

        do {
            return try JSONDecoder().decode(GoogleOAuthTokenResponse.self, from: data)
        } catch {
            throw GoogleOAuthError.tokenEndpointFailed("Google token response could not be read.")
        }
    }

    private func randomBase64URL(byteCount: Int) throws -> String {
        var bytes = [UInt8](repeating: 0, count: byteCount)
        let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        guard status == errSecSuccess else {
            throw GoogleOAuthKeychainError.unhandledStatus(status)
        }
        return Data(bytes).base64URLEncodedString()
    }

    private func codeChallenge(for verifier: String) -> String {
        let digest = SHA256.hash(data: Data(verifier.utf8))
        return Data(digest).base64URLEncodedString()
    }

    private func percentEncodedForm(_ values: [String: String]) -> String {
        values
            .sorted { $0.key < $1.key }
            .map { "\(formEscape($0.key))=\(formEscape($0.value))" }
            .joined(separator: "&")
    }

    private func formEscape(_ value: String) -> String {
        var allowed = CharacterSet.urlQueryAllowed
        allowed.remove(charactersIn: ":#[]@!$&'()*+,;=")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private static func formEscape(_ value: String) -> String {
        var allowed = CharacterSet.urlQueryAllowed
        allowed.remove(charactersIn: ":#[]@!$&'()*+,;=")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    private static func hasRequiredCalendarScopes(_ scopes: [String]) -> Bool {
        let granted = Set(scopes)
        return calendarScopes.allSatisfy { granted.contains($0) }
    }
}

private struct GoogleOAuthTokenResponse: Decodable {
    let accessToken: String
    let expiresIn: Int
    let refreshToken: String?
    let scope: String?

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case expiresIn = "expires_in"
        case refreshToken = "refresh_token"
        case scope
    }
}

private struct GoogleOAuthErrorResponse: Decodable {
    let error: String?
    let errorDescription: String?

    var requiresStoredCredentialReconnect: Bool {
        guard let error else { return false }
        return error == "invalid_grant" || error == "invalid_scope"
    }

    var safeDescription: String {
        errorDescription ?? error ?? "Google authorization failed."
    }

    enum CodingKeys: String, CodingKey {
        case error
        case errorDescription = "error_description"
    }
}

private extension Data {
    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: windows/src/HoverPocket.Shell/Services/GoogleOAuthService.cs
using System.Diagnostics;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HoverPocket.Shell.Services;

internal sealed record GoogleOAuthConfiguration(
    string ClientId,
    string? ClientSecret)
{
    public static string AppDataRoot =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "HoverPocket");

    public static string ConfigurationPath => Path.Combine(AppDataRoot, "oauth.json");

    public static GoogleOAuthConfiguration? Load()
    {
        return LoadFromSources(LoadEmbedded(), ConfigurationPath);
    }

    internal static GoogleOAuthConfiguration? LoadEmbedded()
    {
        var metadata = Assembly.GetExecutingAssembly()
            .GetCustomAttributes<AssemblyMetadataAttribute>()
            .ToArray();
        return FromValues(
            ReadMetadata(metadata, "GoogleOAuthClientId"),
            ReadMetadata(metadata, "GoogleOAuthClientSecret"));
    }

    internal static GoogleOAuthConfiguration? LoadFromSources(
        GoogleOAuthConfiguration? embeddedConfiguration,
        string configurationPath)
    {
        return embeddedConfiguration ?? LoadFromJsonFile(configurationPath);
    }

    internal static GoogleOAuthConfiguration? FromValues(string? clientId, string? clientSecret)
    {
        var normalizedClientId = clientId?.Trim();
        if (string.IsNullOrWhiteSpace(normalizedClientId))
        {
            return null;
        }

        var normalizedClientSecret = clientSecret?.Trim();
        return new GoogleOAuthConfiguration(
            normalizedClientId,
            string.IsNullOrWhiteSpace(normalizedClientSecret) ? null : normalizedClientSecret);
    }

    private static GoogleOAuthConfiguration? LoadFromJsonFile(string configurationPath)
    {
        if (!File.Exists(configurationPath))
        {
            return null;
        }

        try
        {
            using var document = JsonDocument.Parse(File.ReadAllText(configurationPath));
            var root = document.RootElement;
            var container = root.TryGetProperty("installed", out var installed) ? installed : root;
            var clientId = ReadString(container, "client_id") ?? ReadString(root, "clientId");
            var clientSecret = ReadString(container, "client_secret") ?? ReadString(root, "clientSecret");
            return FromValues(clientId, clientSecret);
        }
        catch (JsonException)
        {
            return null;
        }
        catch (IOException)
        {
            return null;
        }
        catch (UnauthorizedAccessException)
        {
            return null;
        }
    }

    private static string? ReadMetadata(IEnumerable<AssemblyMetadataAttribute> metadata, string key)
    {
        return metadata
            .Where(item => string.Equals(item.Key, key, StringComparison.Ordinal))
            .Select(item => item.Value)
            .FirstOrDefault(value => !string.IsNullOrWhiteSpace(value));
    }

    private static string? ReadString(JsonElement element, string propertyName)
    {
        return element.TryGetProperty(propertyName, out var property) && property.ValueKind == JsonValueKind.String
            ? property.GetString()?.Trim()
            : null;
    }
}

internal sealed record GoogleOAuthToken(
    string AccessToken,
    DateTimeOffset ExpiresAt,
    IReadOnlyList<string> GrantedScopes)
{
    public bool IsFresh => ExpiresAt > DateTimeOffset.UtcNow.AddSeconds(60);
}

internal enum GoogleOAuthStoredCredentialStatus
{
    Missing,
    NeedsReconnect,
    Ready
}

internal sealed class GoogleOAuthException : Exception
{
    public GoogleOAuthException(string code, string message, bool requiresReconnect = false)
        : base(message)
    {
        Code = code;
        RequiresReconnect = requiresReconnect;
    }

    public string Code { get; }

    public bool RequiresReconnect { get; }
}

internal sealed record GoogleOAuthAuthorizationRequest(
    Uri Url,
    string State,
    string CodeVerifier,
    string CodeChallenge,
    string RedirectUri);

internal sealed class GoogleOAuthService
{
    public const string CalendarScope = "https://www.googleapis.com/auth/calendar";
    public const string CalendarEventsScope = "https://www.googleapis.com/auth/calendar.events";
    public const string CalendarReadonlyScope = "https://www.googleapis.com/auth/calendar.readonly";
    public const string CalendarListScope = "https://www.googleapis.com/auth/calendar.calendarlist";
    public const string CalendarListReadonlyScope = "https://www.googleapis.com/auth/calendar.calendarlist.readonly";

    public static readonly IReadOnlyList<string> CalendarScopes =
    [
        CalendarEventsScope,
        CalendarListReadonlyScope
    ];

    private static readonly HttpClient HttpClient = new();
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) }
    };

    private readonly GoogleOAuthCredentialStore _credentialStore;
    private readonly object _tokenLock = new();
    private GoogleOAuthToken? _currentToken;

    public GoogleOAuthService(GoogleOAuthCredentialStore? credentialStore = null)
    {
        _credentialStore = credentialStore ?? new GoogleOAuthCredentialStore();
    }

    public bool IsConfigured => GoogleOAuthConfiguration.Load() is not null;

    public bool HasStoredCredential() => StoredCredentialStatus() != GoogleOAuthStoredCredentialStatus.Missing;

    public bool HasRequiredCalendarCredential() => StoredCredentialStatus() == GoogleOAuthStoredCredentialStatus.Ready;

    public GoogleOAuthStoredCredentialStatus StoredCredentialStatus()
    {
        GoogleOAuthStoredCredential? credential;
        try
        {
            credential = _credentialStore.Load();
        }
        catch (InvalidOperationException)
        {
            return GoogleOAuthStoredCredentialStatus.NeedsReconnect;
        }

        if (credential is null)
        {
            return GoogleOAuthStoredCredentialStatus.Missing;
        }

        return HasRequiredCalendarScopes(credential.GrantedScopes)
            ? GoogleOAuthStoredCredentialStatus.Ready
            : GoogleOAuthStoredCredentialStatus.NeedsReconnect;
    }

    public async Task SignInAsync(CancellationToken cancellationToken = default)
    {
        var configuration = GoogleOAuthConfiguration.Load()
            ?? throw new GoogleOAuthException("missing_configuration", "Google OAuth client is not configured.");

        using var receiver = new LoopbackOAuthReceiver();
        var authorization = CreateAuthorizationRequest(configuration, receiver.RedirectUri);
        if (!OpenDefaultBrowser(authorization.Url))
        {
            throw new GoogleOAuthException("browser_open_failed", "Could not open the Google sign-in page.");
        }

        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(TimeSpan.FromMinutes(3));
        OAuthCallback callback;
        try
        {
            callback = await receiver.WaitForCallbackAsync(timeout.Token);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            throw new GoogleOAuthException("timed_out", "Google sign-in timed out.");
        }

        if (!string.IsNullOrWhiteSpace(callback.Error))
        {
            throw new GoogleOAuthException("user_denied", callback.Error);
        }

        if (!string.Equals(callback.State, authorization.State, StringComparison.Ordinal))
        {
            throw new GoogleOAuthException("state_mismatch", "Google sign-in state validation failed.");
        }

        if (string.IsNullOrWhiteSpace(callback.Code))
        {
            throw new GoogleOAuthException("missing_code", "Google did not return an authorization code.");
        }

        var response = await ExchangeAuthorizationCodeAsync(
            configuration,
            callback.Code,
            authorization.CodeVerifier,
            authorization.RedirectUri,
            cancellationToken);

        if (string.IsNullOrWhiteSpace(response.RefreshToken))
        {
            throw new GoogleOAuthException("missing_refresh_token", "Google did not return a refresh token.", requiresReconnect: true);
        }

        var scopes = ParseScopes(response.Scope);
        if (!HasRequiredCalendarScopes(scopes))
        {
            throw new GoogleOAuthException("insufficient_scopes", "Reconnect Google Calendar to allow event editing.", requiresReconnect: true);
        }

        _credentialStore.Save(new GoogleOAuthStoredCredential(response.RefreshToken, scopes));
        SetCurrentToken(new GoogleOAuthToken(
            response.AccessToken,
            DateTimeOffset.UtcNow.AddSeconds(response.ExpiresIn),
            scopes));
    }

    public void SignOut()
    {
        SetCurrentToken(null);
        _credentialStore.Delete();
    }

    public async Task<string> AccessTokenAsync(bool forceRefresh = false, CancellationToken cancellationToken = default)
    {
        _ = GoogleOAuthConfiguration.Load()
            ?? throw new GoogleOAuthException("missing_configuration", "Google OAuth client is not configured.");

        if (!forceRefresh && ReadCurrentToken() is { IsFresh: true } token)
        {
            if (!HasRequiredCalendarScopes(token.GrantedScopes))
            {
                throw new GoogleOAuthException("insufficient_scopes", "Reconnect Google Calendar to allow event editing.", requiresReconnect: true);
            }

            return token.AccessToken;
        }

        var stored = _credentialStore.Load()
            ?? throw new GoogleOAuthException("missing_refresh_token", "Google Calendar is not connected.", requiresReconnect: true);

        if (!HasRequiredCalendarScopes(stored.GrantedScopes))
        {
            throw new GoogleOAuthException("insufficient_scopes", "Reconnect Google Calendar to allow event editing.", requiresReconnect: true);
        }

        try
        {
            var response = await RefreshAccessTokenAsync(stored.RefreshToken, cancellationToken);
            var scopes = string.IsNullOrWhiteSpace(response.Scope) ? stored.GrantedScopes : ParseScopes(response.Scope);
            if (!HasRequiredCalendarScopes(scopes))
            {
                throw new GoogleOAuthException("insufficient_scopes", "Reconnect Google Calendar to allow event editing.", requiresReconnect: true);
            }

            var refreshedToken = new GoogleOAuthToken(
                response.AccessToken,
                DateTimeOffset.UtcNow.AddSeconds(response.ExpiresIn),
                scopes);
            SetCurrentToken(refreshedToken);
            return refreshedToken.AccessToken;
        }
        catch (GoogleOAuthException ex) when (ex.RequiresReconnect)
        {
            SetCurrentToken(null);
            _credentialStore.Delete();
            throw;
        }
    }

    internal static GoogleOAuthAuthorizationRequest CreateAuthorizationRequest(
        GoogleOAuthConfiguration configuration,
        string redirectUri)
    {
        var state = RandomBase64Url(32);
        var verifier = RandomBase64Url(64);
        var challenge = CodeChallenge(verifier);
        var query = PercentEncodedForm(new Dictionary<string, string>
        {
            ["client_id"] = configuration.ClientId,
            ["redirect_uri"] = redirectUri,
            ["response_type"] = "code",
            ["scope"] = string.Join(' ', CalendarScopes),
            ["access_type"] = "offline",
            ["prompt"] = "consent",
            ["state"] = state,
            ["code_challenge"] = challenge,
            ["code_challenge_method"] = "S256"
        });
        return new GoogleOAuthAuthorizationRequest(
            new Uri($"https://accounts.google.com/o/oauth2/v2/auth?{query}"),
            state,
            verifier,
            challenge,
            redirectUri);
    }

    internal static bool HasRequiredCalendarScopes(IEnumerable<string> scopes)
    {
        var granted = scopes.ToHashSet(StringComparer.Ordinal);
        return HasCalendarEventsScope(granted) && HasCalendarListScope(granted);
    }

    private static bool HasCalendarEventsScope(ISet<string> granted)
    {
        return granted.Contains(CalendarEventsScope) || granted.Contains(CalendarScope);
    }

    private static bool HasCalendarListScope(ISet<string> granted)
    {
        return granted.Contains(CalendarListReadonlyScope)
            || granted.Contains(CalendarListScope)
            || granted.Contains(CalendarReadonlyScope)
            || granted.Contains(CalendarScope);
    }

    internal static string CodeChallenge(string verifier)
    {
        return Base64Url(SHA256.HashData(Encoding.ASCII.GetBytes(verifier)));
    }

    private async Task<GoogleOAuthTokenResponse> ExchangeAuthorizationCodeAsync(
        GoogleOAuthConfiguration configuration,
        string code,
        string verifier,
        string redirectUri,
        CancellationToken cancellationToken)
    {
        var form = new Dictionary<string, string>
        {
            ["client_id"] = configuration.ClientId,
            ["code"] = code,
            ["code_verifier"] = verifier,
            ["grant_type"] = "authorization_code",
            ["redirect_uri"] = redirectUri
        };
        if (!string.IsNullOrWhiteSpace(configuration.ClientSecret))
        {
            form["client_secret"] = configuration.ClientSecret;
        }

        return await PostTokenRequestAsync(form, cancellationToken);
    }

    private async Task<GoogleOAuthTokenResponse> RefreshAccessTokenAsync(
        string refreshToken,
        CancellationToken cancellationToken)
    {
        var configuration = GoogleOAuthConfiguration.Load()
            ?? throw new GoogleOAuthException("missing_configuration", "Google OAuth client is not configured.");
        var form = new Dictionary<string, string>
        {
            ["client_id"] = configuration.ClientId,
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refreshToken
        };
        if (!string.IsNullOrWhiteSpace(configuration.ClientSecret))
        {
            form["client_secret"] = configuration.ClientSecret;
        }

        return await PostTokenRequestAsync(form, cancellationToken, treatsStoredCredentialFailureAsReconnect: true);
    }

    private static async Task<GoogleOAuthTokenResponse> PostTokenRequestAsync(
        Dictionary<string, string> form,
        CancellationToken cancellationToken,
        bool treatsStoredCredentialFailureAsReconnect = false)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, "https://oauth2.googleapis.com/token");
        request.Content = new StringContent(PercentEncodedForm(form), Encoding.UTF8, "application/x-www-form-urlencoded");
        using var response = await HttpClient.SendAsync(request, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
        {
            var error = TryDeserialize<GoogleOAuthErrorResponse>(body);
            if (treatsStoredCredentialFailureAsReconnect && error?.RequiresReconnect == true)
            {
                throw new GoogleOAuthException("stored_credential_requires_reconnect", "Reconnect Google Calendar to continue.", requiresReconnect: true);
            }

            throw new GoogleOAuthException("token_endpoint_failed", error?.SafeDescription ?? "Google token request failed.");
        }

        var token = TryDeserialize<GoogleOAuthTokenResponse>(body);
        return token ?? throw new GoogleOAuthException("token_decode_failed", "Google token response could not be read.");
    }

    private static T? TryDeserialize<T>(string json)
    {
        try
        {
            return JsonSerializer.Deserialize<T>(json, JsonOptions);
        }
        catch (JsonException)
        {
            return default;
        }
    }

    private static bool OpenDefaultBrowser(Uri url)
    {
        try
        {
            using var _ = Process.Start(new ProcessStartInfo(url.ToString())
            {
                UseShellExecute = true
            });
            return true;
        }
        catch (InvalidOperationException)
        {
            return false;
        }
        catch (System.ComponentModel.Win32Exception)
        {
            return false;
        }
    }

    private GoogleOAuthToken? ReadCurrentToken()
    {
        lock (_tokenLock)
        {
            return _currentToken;
        }
    }

    private void SetCurrentToken(GoogleOAuthToken? token)
    {
        lock (_tokenLock)
        {
            _currentToken = token;
        }
    }

    private static IReadOnlyList<string> ParseScopes(string? scope)
    {
        return string.IsNullOrWhiteSpace(scope)
            ? CalendarScopes
            : scope.Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
    }

    private static string RandomBase64Url(int byteCount)
    {
        var bytes = new byte[byteCount];
        RandomNumberGenerator.Fill(bytes);
        return Base64Url(bytes);
    }

    private static string Base64Url(byte[] bytes)
    {
        return Convert.ToBase64String(bytes)
            .Replace("+", "-", StringComparison.Ordinal)
            .Replace("/", "_", StringComparison.Ordinal)
            .Replace("=", string.Empty, StringComparison.Ordinal);
    }

    private static string PercentEncodedForm(Dictionary<string, string> values)
    {
        return string.Join('&', values
            .OrderBy(pair => pair.Key, StringComparer.Ordinal)
            .Select(pair => $"{Escape(pair.Key)}={Escape(pair.Value)}"));
    }

    private static string Escape(string value)
    {
        return Uri.EscapeDataString(value).Replace("%20", "+", StringComparison.Ordinal);
    }

    private sealed record GoogleOAuthTokenResponse(
        [property: JsonPropertyName("access_token")] string AccessToken,
        [property: JsonPropertyName("expires_in")] int ExpiresIn,
        [property: JsonPropertyName("refresh_token")] string? RefreshToken,
        [property: JsonPropertyName("scope")] string? Scope);

    private sealed record GoogleOAuthErrorResponse(
        [property: JsonPropertyName("error")] string? Error,
        [property: JsonPropertyName("error_description")] string? ErrorDescription)
    {
        public bool RequiresReconnect => Error is "invalid_grant" or "invalid_scope";

        public string SafeDescription => ErrorDescription ?? Error ?? "Google authorization failed.";
    }
}
